File: qa_automation/main_gnb.py
import requests
import csv
import json
import warnings
from bs4 import BeautifulSoup, NavigableString, MarkupResemblesLocatorWarning
from datetime import datetime
# from pandas import ExcelWriter
from openpyxl import Workbook
from openpyxl.styles import PatternFill, Alignment
import pandas as pd
from process_components import *
from data_management import *

# url_codes = ['uk', 'fr', 'de', 'it', 'es', 'sg', 'id', 'ph', 'au', 'my', 'th', 'vn', 'ar', 'mx', 'br', 'in', 'sa', 'sa_en', 'ca', 'ca_fr', 'ae', 'ae_ar']
url_codes = ['fr', 'de', 'es', 'id', 'vn']

warnings.filterwarnings("ignore", category=MarkupResemblesLocatorWarning)

def check_404 (extracted_link):
    if extracted_link[:4].lower() != "http" and extracted_link[:2].lower() != "//":
        url_link = 'https://www.samsung.com' + extracted_link
    elif extracted_link[:2].lower() == "//":
        url_link = 'https:' + extracted_link
    else:
        url_link = extracted_link

    if url_link.lower().endswith('.pdf'):
        try:
            check_response = requests.head(url_link, allow_redirects=True, timeout=5)
            if check_response.status_code == 200:
                link_page_result = 'Y'
            else:
                link_page_result = 'N'
        except check_response.RequestException as e:
            link_page_result = 'N'

    elif 'www.samsung.com/global/' in url_link.lower() :
        check_response = requests.get(url_link)
        check_html_all = BeautifulSoup(check_response.text, 'lxml')
        if check_html_all.select_one('#content > div > div > div > div > div.cm-g-error-page') is not None:
            link_page_result = 'N'
        else:
            link_page_result = 'Y'

    elif 'shop.samsung.com/' in url_link.lower():
        link_page_result = 'N/A'

    elif 'news.samsung.com/' in url_link.lower():
        check_response = requests.get(url_link, allow_redirects=True)
        check_html_all = BeautifulSoup(check_response.text, 'lxml')
        check_text = check_html_all.select_one('head > title').text
        if check_text.lower()[:14] == 'page not found':
            link_page_result = 'N'
        else:
            link_page_result = 'Y'

    elif 'community.samsung.com/' in url_link.lower():
        link_page_result = 'N/A'

    elif 'design.samsung.com/' in url_link.lower():
        link_page_result = 'N/A'

    elif 'csr.samsung.com/' in url_link.lower():
        link_page_result = 'Y'
        check_response = requests.get(url_link)
        check_html_all = BeautifulSoup(check_response.text, 'lxml')
        body_tag = check_html_all.find('body')

        if body_tag and 'errorPage' in body_tag.get('class', []):
            link_page_result = 'N'
        else:
            try:
                p_text = check_html_all.find('p').text
                data = json.loads(p_text)
                if data.get("status") == 404:
                    link_page_result = 'N'
            except (AttributeError, TypeError, json.JSONDecodeError):
                pass

    elif 'www.samsung.com/' in url_link.lower():
        check_response = requests.get(url_link)
        check_html_all = BeautifulSoup(check_response.text, 'lxml')
        if check_html_all.select_one('#content > div > div > div > div > div.cm-g-error-page') is not None:
            link_page_result = 'N'
        else:
            link_page_result = 'Y'

    else:
        link_page_result = 'N/A'


    return link_page_result


def add_item (exl_ws, col_depth1, col_depth2, col_depth3, col_depth4, extracted_link, link_result):
    saved_item = [col_depth1, col_depth2, col_depth3, col_depth4, extracted_link, link_result]
    exl_ws.append(saved_item)

url_codes = ["us"]
for url_code in url_codes:
    url = 'https://www.samsung.com/' + url_code + '/'
    sitecode = url.split('/')[3]
    response = requests.get(url)
    html_all = BeautifulSoup(response.text, 'lxml')

    print(sitecode)



    exl_wb = Workbook()

    exl_ws = exl_wb.active
    exl_ws.title = "Home"

    exl_header = ['Location', 'Main Manu', 'Sub Menu', 'Title', 'URL', 'Check']
    exl_ws.append(exl_header)

    exl_ws.column_dimensions['A'].width = 10
    exl_ws.column_dimensions['B'].width = 20
    exl_ws.column_dimensions['C'].width = 25
    exl_ws.column_dimensions['D'].width = 30
    exl_ws.column_dimensions['E'].width = 50
    exl_ws.column_dimensions['F'].width = 5
    exl_ws.sheet_view.showGridLines = False
    exl_ws.freeze_panes = 'A2'

    content_all = html_all.select('#component-id > div.nv00-gnb__inner-wrap > div.nv00-gnb__l0-menu-wrap > ul')

    col_depth1 = 'GNB'
    for content_menu in content_all:
        col_depth2 = ''
        for content_depth01 in content_menu.select('li.nv00-gnb__l0-menu'):
            if content_depth01.select_one('button.nv00-gnb__l0-menu-btn') is not None:
                content_depth01_title = content_depth01.select_one('button.nv00-gnb__l0-menu-btn')
                print(content_depth01_title.text.strip())
                col_depth2 = content_depth01_title.text.strip()
            elif content_depth01.select_one('a.nv00-gnb__l0-menu-link') is not None:
                content_depth01_title = content_depth01.select_one('a.nv00-gnb__l0-menu-link')
                print(content_depth01_title.text.strip())
                col_depth2 = content_depth01_title.text.strip()
            col_depth3 = ''
            for content_depth02 in content_depth01.select('div.nv00-gnb__l1-menu-wrap'):
                # print(content_depth02)
                if content_depth02.select_one('div > h3.nv00-gnb__l1-menu-btn') is not None:
                    content_depth02_title = content_depth02.select_one('div > h3.nv00-gnb__l1-menu-btn')
                    print (' -' + content_depth02_title.text.strip())
                    col_depth3 = content_depth02_title.text.strip()
                if content_depth02.select_one('p.nv00-gnb__featured-products-thumbnail-title') is not None:
                    content_depth02_title = content_depth02.select_one('p.nv00-gnb__featured-products-thumbnail-title')
                    print (' -' + content_depth02_title.text.strip())
                    col_depth3 = content_depth02_title.text.strip()
                if content_depth02.select('li.nv00-gnb__l1-menu > a') is not None:
                    for content_depth03 in content_depth02.select('li.nv00-gnb__l1-menu > a'):
                        extracted_text = ""
                        for content in content_depth03:
                            if isinstance(content, NavigableString):
                                extracted_text += content
                        extracted_text = extracted_text.strip()
                        extracted_link = content_depth03.get('href', 'None')
                        link_result = check_404(extracted_link)
                        print("    - " + extracted_text, extracted_link, link_result)
                        col_depth4 = extracted_text
                        add_item(exl_ws, col_depth1, col_depth2, col_depth3, col_depth4, extracted_link, link_result)
                if content_depth02.select('li.nv00-gnb__l2-menu > a') is not None:
                    for content_depth03 in content_depth02.select('li.nv00-gnb__l2-menu > a'):

                        extracted_text = ""
                        for content in content_depth03:
                            if isinstance(content, NavigableString):
                                extracted_text += content
                        extracted_text = extracted_text.strip()
                        extracted_link = content_depth03.get('href', 'None')
                        link_result = check_404(extracted_link)
                        print("    - " + extracted_text, extracted_link, link_result)
                        col_depth4 = extracted_text
                        add_item(exl_ws, col_depth1, col_depth2, col_depth3, col_depth4, extracted_link, link_result)
                if content_depth02.select('li.nv00-gnb__featured-products-thumbnail-item > a') is not None:
                    for content_depth03 in content_depth02.select('li.nv00-gnb__featured-products-thumbnail-item > a'):
                        extracted_text = content_depth03.select_one('p.nv00-gnb__featured-products-thumbnail-item-name')
                        extracted_link = content_depth03.get('href', 'None')
                        link_result = check_404(extracted_link)
                        print("    - " + extracted_text.text.strip(), extracted_link, link_result)
                        col_depth4 = extracted_text.text.strip()
                        add_item(exl_ws, col_depth1, col_depth2, col_depth3, col_depth4, extracted_link, link_result)

    col_depth1 = 'Footer'
    col_depth2 = ''
    content_all = html_all.select('footer > div.footer > div.footer-column > div.footer-column__item')
    for content_menu in content_all:
        content_depth01_title = content_menu.select_one('div > h3.footer-category__title')
        col_depth3 = content_depth01_title.text.strip()
        print(content_depth01_title.text.strip())
        for content_depth02 in content_menu.select('div > div.footer-category__list-wrap > ul > li.footer-category__item'):
            content_depth02_title = content_depth02.select_one('a.footer-category__link')
            content_depth02_link = content_depth02_title.get('href', 'None')
            link_result = check_404(content_depth02_link)
            col_depth4 = content_depth02_title.text.strip()
            print(" - " + content_depth02_title.text.strip() , content_depth02_link, link_result)
            add_item(exl_ws, col_depth1, col_depth2, col_depth3, col_depth4, content_depth02_link, link_result)

    df = pd.DataFrame(data_list)
    file_name_base = url.replace('https://www.samsung.com/', '').replace('/', '_')
    file_name_base = file_name_base.strip('-')
    current_time = datetime.now().strftime("%Y%m%d-%H%M%S")
    final_file_name = f"{file_name_base}_{current_time}.xlsx"

    exl_wb.save(final_file_name)

File: qa_automation/jsrefactor/main_gnb.js
import axios from "axios";
import * as cheerio from "cheerio";
import exceljs from "exceljs";
const { Workbook } = exceljs;

// const url_codes = ['uk', 'fr', 'de', 'it', 'es', 'sg', 'id', 'ph', 'au', 'my', 'th', 'vn', 'ar', 'mx', 'br', 'in', 'sa', 'sa_en', 'ca', 'ca_fr', 'ae', 'ae_ar']
// const url_codes = ["fr", "de", "es", "id", "vn"];
const url_codes = ["fr"];

// BS에서 파싱 패턴에서 HTML 주석처럼 보일 수 있는 에러 출력 방지
// warnings.filterwarnings("ignore", category=MarkupResemblesLocatorWarning)

const check_404 = async (extracted_link) => {
  const url_link = extracted_link;
  let link_page_result = "N";

  // URL이 내부 인지 (ex .com)
  if (
    extracted_link.slice(0, 4).toLowerCase() !== "http" &&
    extracted_link.slice(0, 2).toLowerCase() !== "//"
  ) {
    url_link = "https://www.samsung.com" + extracted_link;
  }

  // 로 시작하는 ex) 이미지 서버
  else if (extracted_link.slice(0, 2).toLowerCase() == "//") {
    url_link = "https:" + extracted_link;
  }

  // 파일로 연결되는 경우 (ex. pdf)
  if (url_link.toLowerCase().endsWith(".pdf")) {
    try {
      const check_response = await axios.head(url_link, {
        allowRedirects: true,
        timeout: 5000,
      });
      if (check_response.status === 200) {
        link_page_result = "Y";
      } else {
        link_page_result = "N";
      }
    } catch (e) {
      // 굳이?
      if (e.response) {
        console.error("Error response from server(.pdf):", e.response.data);
      } else if (e.request) {
        console.error("Request failed(.pdf):", e.request);
      } else {
        console.error("Error(.pdf):", e.message);
      }
      link_page_result = "N";
    }
  }

  // 글로벌 링크로 연결시
  else if (url_link.toLowerCase().includes("www.samsung.com/global/")) {
    try {
      const { data } = await axios.get(url_link);
      const $check_html_all = cheerio.load(data);
      if (
        $check_html_all(
          "#content > div > div > div > div > div.cm-g-error-page"
        ).length > 0
      ) {
        link_page_result = "N";
      } else {
        link_page_result = "Y";
      }
    } catch (e) {
      if (e.response) {
        console.error(
          "Error response from server:(www.samsung.com/global/)",
          e.response.data
        );
      } else if (e.request) {
        console.error("Request failed:(www.samsung.com/global/)", e.request);
      } else {
        console.error("Error:(www.samsung.com/global/)", e.message);
      }
      link_page_result = "N";
    }
  } else if (url_link.toLowerCase().includes("shop.samsung.com/")) {
    link_page_result = "N/A";
  } else if (url_link.toLowerCase().includes("news.samsung.com/")) {
    try {
      const { data } = await axios.get(url_link, { allowRedirects: true });
      const $check_html_all = cheerio.load(data, { xmlMode: true });
      const check_text = $check_html_all("head > title").text();
      if (check_text.toLowerCase().slice(0, 14) === "page not found") {
        link_page_result = "N";
      } else {
        link_page_result = "Y";
      }
    } catch (e) {
      if (e.response) {
        console.error(
          "Error response from server(news.samsung.com/):",
          e.response.data
        );
      } else if (e.request) {
        console.error("Request failed(news.samsung.com/):", e.request);
      } else {
        console.error("Error(news.samsung.com/):", e.message);
      }
      link_page_result = "N";
    }
  } else if (url_link.toLowerCase().includes("community.samsung.com/")) {
    link_page_result = "N";
  } else if (url_link.toLowerCase().includes("design.samsung.com/")) {
    link_page_result = "N";
  } else if (url_link.toLowerCase().includes("csr.samsung.com/")) {
    link_page_result = "Y";
    const { data } = axios.get(url_link);
    const $check_html_all = cheerio.load(data, { xmlMode: true });
    const body_tag = $check_html_all("body");

    if (
      body_tag &&
      body_tag.attr("class") &&
      body_tag.attr("class").includes("errorPage")
    ) {
      link_page_result = "N";
    } else {
      try {
        const p_text = $check_html_all("p").text();
        const data = JSON.parse(p_text);
        if (data.status === 404) {
          link_page_result = "N";
        }
      } catch (error) {
        // node에선 AttrError가 타입에러로 발생함
        if (
          error instanceof TypeError ||
          error instanceof json.JSONDecodeError
        ) {
          // TypeError 또는 JSON 파싱 오류가 발생한 경우 PASS
        } else {
          console.error("Error(csr.samsung.com/):", error.message);
        }
      }
    }
  } else if (url_link.toLowerCase().includes("www.samsung.com/")) {
    try {
      const { data } = axios.get(url_link);
      const $check_html_all = cheerio.load(data);
      const errorPageElement = $check_html_all(
        "#content > div > div > div > div > div.cm-g-error-page"
      );
      if (errorPageElement.length > 0) {
        link_page_result = "N";
      } else {
        link_page_result = "Y";
      }
    } catch (e) {
      console.error("Error(www.samsung.com/):", error.message);
    }
  } else link_page_result = "N/A";

  return link_page_result;
};

const add_item = async (
  exl_ws,
  col_depth1,
  col_depth2,
  col_depth3,
  col_depth4,
  extracted_link,
  link_result
) => {
  const saved_item = [
    col_depth1,
    col_depth2,
    col_depth3,
    col_depth4,
    extracted_link,
    link_result,
  ];
  exl_ws.addRow(saved_item);
};

url_codes.forEach(async (url_code, index) => {
  const url = `https://www.samsung.com/${url_code}/`;
  const sitecode = url.split("/")[3];

  try {
    // Excel 워크북 생성
    const exl_wb = new Workbook();

    // 활성 시트 생성 및 이름 지정
    const exl_ws = exl_wb.addWorksheet("Home");

    // 헤더 데이터
    const exl_header = [
      "Location",
      "Main Menu",
      "Sub Menu",
      "Title",
      "URL",
      "Check",
    ];

    // 헤더 추가
    exl_ws.addRow(exl_header);

    // 열의 너비 설정
    exl_ws.columns = [
      { width: 10 }, // A열
      { width: 20 }, // B열
      { width: 25 }, // C열
      { width: 30 }, // D열
      { width: 50 }, // E열
      { width: 5 }, // F열
    ];

    // 그리드 라인 숨기기, A2 기준으로 고정
    exl_ws.views = [
      {
        showGridLines: false,
        state: "frozen",
        xSplit: 0,
        ySplit: 1,
        topLeftCell: "A2",
      },
    ];

    const { data } = await axios.get(url);
    console.log(sitecode);
    const $ = cheerio.load(data, { xmlMode: true });
    const content_all = [];
    $(
      "#component-id > div.nv00-gnb__inner-wrap > div.nv00-gnb__l0-menu-wrap > ul"
    ).each((_, el) => {
      content_all.push(el);
    });

    let col_depth1 = "GNB";
    // Going to depth 1
    content_all.forEach((content_menu, _idx) => {
      const $ = cheerio.load(content_menu, { xmlMode: true });
      content_menu = [];
      $("li.nv00-gnb__l0-menu").each((_, el) => {
        content_menu.push(el);
      });

      let col_depth2 = "";
      // Going to depth 2
      content_menu.forEach((content_depth01, _idx) => {
        const $ = cheerio.load(content_depth01, { xmlMode: true });
        let content_depth01_title = $("button.nv00-gnb__l0-menu-btn");
        if (content_depth01_title.html() !== null) {
          content_depth01_title = content_depth01_title.first();
          console.log(content_depth01_title.text().trim());
          col_depth2 = content_depth01_title.text().trim();
        }
        content_depth01_title = $("a.nv00-gnb__l0-menu-link");
        if (content_depth01_title.html() !== null) {
          content_depth01_title = content_depth01_title.first();
          console.log(content_depth01_title.text().trim());
          col_depth2 = content_depth01_title.text().trim();
        }

        // Going to depth 3
        let col_depth3 = "";
        content_depth01 = [];
        $("div.nv00-gnb__l1-menu-wrap").each((_, el) => {
          content_depth01.push(el);
        });
        content_depth01.forEach((content_depth02, _idx) => {
          const $ = cheerio.load(content_depth02, { xmlMode: true });
          let content_depth02_title = $(
            "div > h3.nv00-gnb__l1-menu-btn"
          ).first();
          if (content_depth02_title.html() !== null) {
            console.log("-", content_depth02_title.text().trim());
            col_depth3 = content_depth02_title.text().trim();
          }

          content_depth02_title = $(
            "p.nv00-gnb__featured-products-thumbnail-title"
          ).first();
          if (content_depth02_title.html() !== null) {
            console.log("-", content_depth02_title.text().trim());
            col_depth3 = content_depth02_title.text().trim();
          }

          content_depth02 = [];
          $("li.nv00-gnb__l1-menu > a").each((_, el) => {
            content_depth02.push(el);
          });
          // MOBIE의 좌측 같이 하위 네비게이션이 없는 경우
          if (content_depth02.length > 0) {
            console.log("@@@@@@");
            content_depth02.forEach((content_depth03, _idx) => {
              let extracted_text = "";
              console.log("@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@ \n");
              console.log(content_depth03);
            });
          }
        });
      });
    });
  } catch (error) {
    console.error("Error:", error.message);
  }
});
